### DISCUSSION TITLE
Make (someIntPtr is null) work

### DISCUSSION THREAD

**Discussion by GSPP** [dotnet/csharplang#5764] [General]
Testing an `IntPtr` for `null` requires comparing to `IntPtr.Zero`. While this is OK, it could be simpler:

```
IntPtr someIntPtr;

if (someIntPtr is null) { ... } //1
if (someIntPtr == null) { ... } //2
```

Can this be made to work?

(2) actually works if you don't use C# strict mode (see https://github.com/dotnet/roslyn/issues/18975). So it's not supposed to work and you're not supposed to do that. It would be nice if this was supported.

**Comment by RikkiGibson** [comment]
Moving to csharplang as this appears to be a request to change the language.

**Reply by GSPP** [reply]
Thanks.

**Comment by tannergooding** [comment]
IntPtr is not a pointer. It is a pointer-bit integer (in the same way Int32 is a 32-bit integer). It's name actually comes from C/C++ where there is `intptr_t`

Because it is a pointer sized integer, it is sometimes used to hold or otherwise manipulate pointers, addresses, or related data.

However, it is ultimately still an integer just as `int` and `long` are and the comparison against `null` is typically a bug in these cases (and was/is certainly a common source of bugs in C/C++).

It would not be, IMO, good to allow a comparison to null here. Instead, you're likely better off using `nint` and directly comparing against the literal `0`. This will also result in smaller IL and ultimately better JIT throughput

**Comment by 333fred** [comment]
2 "works", but does not actually do what you think it does. It does not compare against `IntPtr.Zero`, it compares against `null` the reference: this is always false. In general, I don't think we are interested in treating `IntPtr` or `UIntPtr` different from any other struct type: they cannot be null.
